# Student Loan Risk with Deep Learning

In [ ]:
# Imports
import pandas as pd
import tensorflow as tf
from tensorflow.keras.layers import Dense
from tensorflow.keras.models import Sequential
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import classification_report
from pathlib import Path

---

## Prepare the data to be used on a neural network model

### Step 1: Read the `student-loans.csv` file into a Pandas DataFrame. Review the DataFrame, looking for columns that could eventually define your features and target variables.   

In [ ]:
# Read the csv into a Pandas DataFrame
file_path = "https://static.bc-edx.com/ai/ail-v-1-0/m18/lms/datasets/student-loans.csv"
loans_df = pd.read_csv(file_path)

# Review the DataFrame
loans_df.head()

payment_history  location_parameter  stem_degree_score  gpa_ranking  \
0              7.4                0.70               0.00          1.9   
1              7.8                0.88               0.00          2.6   
2              7.8                0.76               0.04          2.3   
3             11.2                0.28               0.56          1.9   
4              7.4                0.70               0.00          1.9   

   alumni_success  study_major_code  time_to_completion  \
0           0.076              11.0                34.0   
1           0.098              25.0                67.0   
2           0.092              15.0                54.0   
3           0.075              17.0                60.0   
4           0.076              11.0                34.0   

   finance_workshop_score  cohort_ranking  total_loan_score  \
0                  0.9978            3.51              0.56   
1                  0.9968            3.20              0.68   
2                  0.9970            3.26              0.65   
3                  0.9980            3.16              0.58   
4                  0.9978            3.51              0.56   

   financial_aid_score  credit_ranking  
0                  9.4               0  
1                  9.8               0  
2                  9.8               0  
3                  9.8               1  
4                  9.4               0

In [ ]:
# Review the data types associated with the columns
loans_df.dtypes

payment_history           float64
location_parameter        float64
stem_degree_score         float64
gpa_ranking               float64
alumni_success            float64
study_major_code          float64
time_to_completion        float64
finance_workshop_score    float64
cohort_ranking            float64
total_loan_score          float64
financial_aid_score       float64
credit_ranking              int64
dtype: object

In [ ]:
# Check the credit_ranking value counts
loans_df["credit_ranking"].value_counts()

credit_ranking
1    855
0    744
Name: count, dtype: int64

### Step 2: Using the preprocessed data, create the features (`X`) and target (`y`) datasets. The target dataset should be defined by the preprocessed DataFrame column “credit_ranking”. The remaining columns should define the features dataset.

In [ ]:
# Define the target set y using the credit_ranking column
y = loans_df["credit_ranking"]

# Display a sample of y
y[:5]

0    0
1    0
2    0
3    1
4    0
Name: credit_ranking, dtype: int64

In [ ]:
# Define features set X by selecting all columns but credit_ranking
x = loans_df.drop(columns=["credit_ranking"])


# Review the features DataFrame
x.head()

payment_history  location_parameter  stem_degree_score  gpa_ranking  \
0              7.4                0.70               0.00          1.9   
1              7.8                0.88               0.00          2.6   
2              7.8                0.76               0.04          2.3   
3             11.2                0.28               0.56          1.9   
4              7.4                0.70               0.00          1.9   

   alumni_success  study_major_code  time_to_completion  \
0           0.076              11.0                34.0   
1           0.098              25.0                67.0   
2           0.092              15.0                54.0   
3           0.075              17.0                60.0   
4           0.076              11.0                34.0   

   finance_workshop_score  cohort_ranking  total_loan_score  \
0                  0.9978            3.51              0.56   
1                  0.9968            3.20              0.68   
2                  0.9970            3.26              0.65   
3                  0.9980            3.16              0.58   
4                  0.9978            3.51              0.56   

   financial_aid_score  
0                  9.4  
1                  9.8  
2                  9.8  
3                  9.8  
4                  9.4

### Step 3: Split the features and target sets into training and testing datasets.


In [ ]:
# Split the preprocessed data into a training and testing dataset
# Assign the function a random_state equal to 1
X_train, X_test, y_train, y_test = train_test_split(x, y, random_state=1)

### Step 4: Use scikit-learn's `StandardScaler` to scale the features data.

In [ ]:
# Create a StandardScaler instance
scaler = StandardScaler()

# Fit the scaler to the features training dataset
X_scaler = scaler.fit(X_train)

# Fit the scaler to the features training dataset


---

## Compile and Evaluate a Model Using a Neural Network

### Step 1: Create a deep neural network by assigning the number of input features, the number of layers, and the number of neurons on each layer using Tensorflow’s Keras.

> **Hint** You can start with a two-layer deep neural network model that uses the `relu` activation function for both layers.


In [ ]:
# Define the the number of inputs (features) to the model
number_input_features = len(X_train.iloc[0])

# Review the number of features
number_input_features

11

In [ ]:
# Define the number of hidden nodes for the first hidden layer
first_hidden_nodes = 8

# Define the number of hidden nodes for the second hidden layer
second_hidden_nodes = 5

# Define the number of neurons in the output layer
number_output_neurons = 1

In [ ]:
# Create the Sequential model instance
nn = Sequential()

# Add the first hidden layer
nn.add(Dense(units=first_hidden_nodes, input_dim=number_input_features, activation="relu"))

# Add the second hidden layer
nn.add(Dense(units=second_hidden_nodes, activation="relu"))

# Add the output layer to the model specifying the number of output neurons and activation function
nn.add(Dense(units=number_output_neurons, activation="sigmoid"))


In [ ]:
# Display the Sequential model summary
nn.summary()

Model: "sequential_1"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_3 (Dense)             (None, 8)                 96        
                                                                 
 dense_4 (Dense)             (None, 5)                 45        
                                                                 
 dense_5 (Dense)             (None, 1)                 6         
                                                                 
Total params: 147 (588.00 Byte)
Trainable params: 147 (588.00 Byte)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________


### Step 2: Compile and fit the model using the `binary_crossentropy` loss function, the `adam` optimizer, and the `accuracy` evaluation metric.


In [ ]:
# Compile the Sequential model
nn.compile(loss="binary_crossentropy", optimizer="adam", metrics=["accuracy"])


In [ ]:
# Fit the model using 50 epochs and the training data
model = nn.fit(X_train, y_train, epochs=50)

Epoch 1/50
38/38 [==============================] - 3s 6ms/step - loss: 3.9664 - accuracy: 0.5363
Epoch 2/50
38/38 [==============================] - 0s 6ms/step - loss: 1.7473 - accuracy: 0.4946
Epoch 3/50
38/38 [==============================] - 0s 6ms/step - loss: 0.8724 - accuracy: 0.4621
Epoch 4/50
38/38 [==============================] - 0s 7ms/step - loss: 0.7705 - accuracy: 0.5121
Epoch 5/50
38/38 [==============================] - 0s 8ms/step - loss: 0.7429 - accuracy: 0.5379
Epoch 6/50
38/38 [==============================] - 0s 8ms/step - loss: 0.7242 - accuracy: 0.5488
Epoch 7/50
38/38 [==============================] - 0s 8ms/step - loss: 0.7100 - accuracy: 0.5721
Epoch 8/50
38/38 [==============================] - 0s 9ms/step - loss: 0.6987 - accuracy: 0.5830
Epoch 9/50
38/38 [==============================] - 0s 8ms/step - loss: 0.6904 - accuracy: 0.5905
Epoch 10/50
38/38 [==============================] - 0s 7ms/step - loss: 0.6843 - accuracy: 0.5997
Epoch 11/50
38/38 [

### Step 3: Evaluate the model using the test data to determine the model’s loss and accuracy.


In [ ]:
# Evaluate the model loss and accuracy metrics using the evaluate method and the test data
model_loss, model_accuracy = nn.evaluate(X_test, y_test, verbose=2)

# Display the model loss and accuracy results
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

13/13 - 0s - 3ms/step - accuracy: 0.7475 - loss: 0.5585
Loss: 0.5584709048271179, Accuracy: 0.7475000023841858


### Step 4: Save and export your model to a keras file, and name the file `student_loans.keras`.


In [ ]:
# Set the model's file path

file_path = Path("student_loans.keras")
# Export your model to a keras file
nn.save(file_path)



---
## Predict Loan Repayment Success by Using your Neural Network Model

### Step 1: Reload your saved model.

In [33]:
# Set the model's file path
file_path = Path("student_loans.keras")

# Load the model to a new object
nn_imported = tf.keras.models.load_model(file_path)

### Step 2: Make predictions on the testing data and save the predictions to a DataFrame.

In [ ]:
# Make predictions with the test data
y_predict = nn_imported.predict(X_test)

# Display a sample of the predictions
y_predict[:5]

13/13 [==============================] - 0s 2ms/step


array([[0.554022  ],
       [0.40241408],
       [0.6862471 ],
       [0.65096724],
       [0.831636  ]], dtype=float32)

In [ ]:
# Save the predictions to a DataFrame and round the predictions to binary results
results = pd.DataFrame(y_test)
results["Predicted"] = y_predict
results["Predicted"] = results["Predicted"].apply(lambda x: round(x,2))
results.head(400)


credit_ranking  Predicted
75                 0       0.55
1283               1       0.40
408                1       0.69
1281               1       0.65
1118               1       0.83
...              ...        ...
1416               0       0.65
1567               0       0.69
1204               1       0.70
12                 0       0.45
584                1       0.73

[400 rows x 2 columns]

In [ ]:
# Convert continuous predictions to binary values using a threshold
threshold = 0.5
binary_predictions = [1 if pred >= threshold else 0 for pred in results["Predicted"]]

# Print the classification report with the binary predictions
print(classification_report(results["credit_ranking"], binary_predictions))


              precision    recall  f1-score   support

           0       0.70      0.41      0.52       188
           1       0.62      0.84      0.71       212

    accuracy                           0.64       400
   macro avg       0.66      0.63      0.61       400
weighted avg       0.66      0.64      0.62       400



### Step 4: Display a classification report with the y test data and predictions

In [ ]:
# Convert continuous predictions to binary values using a threshold
threshold = 0.5
binary_predictions = [1 if pred >= threshold else 0 for pred in results["Predicted"]]

# Print the classification report with the binary predictions
print(classification_report(results["credit_ranking"], binary_predictions))


              precision    recall  f1-score   support

           0       0.70      0.41      0.52       188
           1       0.62      0.84      0.71       212

    accuracy                           0.64       400
   macro avg       0.66      0.63      0.61       400
weighted avg       0.66      0.64      0.62       400



---
## Discuss creating a recommendation system for student loans

Briefly answer the following questions in the space provided:

1. Describe the data that you would need to collect to build a recommendation system to recommend student loan options for students. Explain why this data would be relevant and appropriate.

2. Based on the data you chose to use in this recommendation system, would your model be using collaborative filtering, content-based filtering, or context-based filtering? Justify why the data you selected would be suitable for your choice of filtering method.

3. Describe two real-world challenges that you would take into consideration while building a recommendation system for student loans. Explain why these challenges would be of concern for a student loan recommendation system.

**1. Describe the data that you would need to collect to build a recommendation system to recommend student loan options for students. Explain why this data would be relevant and appropriate.**

1.1 Student Demographics, possibly Age, gender, and marital status to helpin understanding the student's background and potentially their financial independence or reliance on family support.

1.2 Academic Information such as type of degree (undergraduate, graduate), institution type (public, private, community college), and academic performance. to determine costs (loan amount) and potential future earnings, which can influence loan options.

1.3 Current Student's Financial Status such as Income, employment status, and existing debts. for assessing the student's histor & ability to repay loans.
   
1.4 Financial Aid Received suchScholarships, grants, and previously taken student loans. This information is necessary to calculate the remaining financial need.

1.5 Future expected salary after graduation, based on field of study and academic performance. This helps in recommending loans that the student can realistically repay based on their potential earnings.

1.6 Historical Data on Loan Performance such as default rates, average repayment times, and satisfaction scores for different loan options. This can help in recommending loans that have a good track record.

1.7 This additional data is relevant and appropriate because it allows the recommendation system to consider a comprehensive view of the student's financial situation, academic background, and future potential. By analyzing this information, the system can recommend loan options that are affordable, suitable for the student's financial and academic profile, and aligned with their preferences and future earning potential.


**2. Based on the data you chose to use in this recommendation system, would your model be using collaborative filtering, content-based filtering, or context-based filtering? Justify why the data you selected would be suitable for your choice of filtering method.**

2.1 The model would primarily use ontent-based filtering. This decision is based on the nature of the data collected, which includes detailed information about the students (demographics, academic information, financial status, loan preferences, and future earnings potential) and the characteristics of different loan options.

2.2 Justification for Content-Based Filtering:

2.2.1. Personalized Recommendation based on data which includes detailed individual profiles, such as academic performance, financial need, and future earnings potential. Content-based filtering can leverage this information to recommend loan options that closely match the student's specific circumstances and preferences.

2.2.2. Direct Attributes Matching as the system can analyze the attributes of loan options such as interest rates, repayment terms and match them with the student's preferences and financial capability. This direct matching of content (loan attributes) with user preferences (student's financial and academic profile) is the essence of content-based filtering.

2.2.3 Focus on Item Features, the recommendation is based on the features of the loans themselves, such as the type of loan, interest rates, and terms of repayment, which can be directly compared to the student's needs and future earning potential. This approach aligns well with content-based filtering, which recommends items based on their features and how these features match the user's profile.

In summary, the detailed, attribute-rich data collected for this recommendation system lends itself well to content-based filtering, where recommendations are made by matching the specific needs and preferences of the student with the characteristics of various loan options. This method allows for personalized, relevant, and effective loan recommendations for each student.


**3. Describe two real-world challenges that you would take into consideration while building a recommendation system for student loans. Explain why these challenges would be of concern for a student loan recommendation system.**

